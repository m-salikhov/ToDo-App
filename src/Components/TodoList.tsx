import { Todo } from './TodoApp';
import TodoItem from './TodoItem';

interface Props {
  todos: Todo[];
  setTodos: React.Dispatch<React.SetStateAction<Todo[]>>;
}
export default function TodoList({ todos, setTodos }: Props) {
  return (
    <>
      {todos.length > 0 ? (
        <ul>
          {todos.map((todo) => {
            if (todo.show) {
              return <TodoItem key={todo.id} todo={todo} setTodos={setTodos} />;
            } else {
              return null;
            }
          })}
        </ul>
      ) : (
        <h4>Todo list is empty</h4>
      )}
    </>
  );
}
