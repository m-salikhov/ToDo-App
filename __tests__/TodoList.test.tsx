import { render, screen } from '@testing-library/react';
import TodoList from '../src/Components/TodoList';

const todos = [
  {
    id: 1,
    text: 'todo 1',
    completed: false,
    show: true,
  },

  {
    id: 2,
    text: 'todo 2',
    completed: true,
    show: true,
  },

  {
    id: 3,
    text: 'todo 3',
    completed: false,
    show: false,
  },
  {
    id: 4,
    text: 'todo 4',
    completed: true,
    show: false,
  },
];

describe('TodoList render', () => {
  test('render empty todo list', () => {
    render(<TodoList todos={[]} setTodos={() => {}} />);

    const fillerElement = screen.getByText('Todo list is empty');
    expect(fillerElement).toBeInTheDocument();
  });

  test('render todo list', () => {
    render(<TodoList todos={todos} setTodos={() => {}} />);

    const list = screen.getByRole('list');
    expect(list).toMatchSnapshot();

    const listItems = screen.getAllByRole('listitem');
    expect(listItems).toHaveLength(2);

    const paragraphDefaultTodo = screen.getByText(/todo 1/i);
    expect(paragraphDefaultTodo).not.toHaveClass('completed');
    expect(paragraphDefaultTodo).toBeInTheDocument();

    const paragraphCompletedTodo = screen.getByText(/todo 2/i);
    expect(paragraphCompletedTodo).toHaveClass('completed');
  });
});

describe('Todo Items render', () => {
  test('uncompleted todo item', () => {
    render(<TodoList todos={todos} setTodos={() => {}} />);

    const paragraphTodo = screen.getByText(/todo 1/i);

    expect(paragraphTodo).toBeInTheDocument();
    expect(paragraphTodo).not.toHaveClass('completed');
  });

  test('completed todo item', () => {
    render(<TodoList todos={todos} setTodos={() => {}} />);

    const paragraphTodo = screen.getByText(/todo 2/i);

    expect(paragraphTodo).toBeInTheDocument();
    expect(paragraphTodo).toHaveClass('completed');
  });
});
