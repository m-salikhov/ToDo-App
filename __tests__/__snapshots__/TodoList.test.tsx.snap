// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList render > render todo list 1`] = `
<ul>
  <li>
    <div
      class="todo-item-icon"
    >
      <div
        class="todo-item-circle"
        data-testid="change-status-icon"
      />
    </div>
    <p>
      todo 1
    </p>
    <div
      class="todo-item-remove"
    >
      <img
        alt="remove todo icon"
        data-testid="delete-icon"
        src="/src/assets/red_cross.svg"
      />
    </div>
  </li>
  <li>
    <div
      class="todo-item-icon"
    >
      <div
        class="todo-item-circle"
        data-testid="change-status-icon"
      >
        <img
          alt="todo completed icon"
          src="/src/assets/green_mark.svg"
        />
      </div>
    </div>
    <p
      class="completed"
    >
      todo 2
    </p>
    <div
      class="todo-item-remove"
    >
      <img
        alt="remove todo icon"
        data-testid="delete-icon"
        src="/src/assets/red_cross.svg"
      />
    </div>
  </li>
</ul>
`;
